pkg/sql/schemachanger/scplan/internal/scstage/build.go

```diff
@@ -434,58 +443,55 @@ func (sb stageBuilder) build() Stage {
 // may prove to be a somewhat common pattern in other cases: consider the
 // intermediate index needed when adding and dropping columns as part of the
 // same transaction.
-func (sb stageBuilder) computeExtraJobOps(s Stage, after []scpb.Status) []scop.Op {
+func (bc buildContext) computeExtraJobOps(s, next *Stage) []scop.Op {
+	revertible := next != nil && next.Phase < scop.PostCommitNonRevertiblePhase
 	switch s.Phase {
 	case scop.PreCommitPhase:
 		// If this pre-commit stage is non-terminal, this means there will be at
 		// least one post-commit stage, so we need to create a schema changer job
 		// and update references for the affected descriptors.
-		if !sb.bc.isStateTerminal(after) {
+		if next != nil {
 			const initialize = true
-			return append(sb.bc.setJobStateOnDescriptorOps(initialize, after),
-				sb.bc.createSchemaChangeJobOp())
+			return append(bc.setJobStateOnDescriptorOps(initialize, revertible, s.After),
+				bc.createSchemaChangeJobOp(revertible))
 		}
 		return nil
 	case scop.PostCommitPhase, scop.PostCommitNonRevertiblePhase:
-		if sb.opType != scop.MutationType {
+		if s.Type() != scop.MutationType {
 			return nil
 		}
 		var ops []scop.Op
-		if sb.bc.isStateTerminal(after) {
+		if next == nil {
 			// The terminal mutation stage needs to remove references to the schema
 			// changer job in the affected descriptors.
-			ops = sb.bc.removeJobReferenceOps()
+			ops = bc.removeJobReferenceOps()
 		} else {
 			const initialize = false
-			ops = sb.bc.setJobStateOnDescriptorOps(initialize, after)
+			ops = bc.setJobStateOnDescriptorOps(initialize, revertible, s.After)
 		}
 		// If we just moved to a non-cancelable phase, we need to tell the job
 		// that it cannot be canceled. Ideally we'd do this just once.
-		//
-		// TODO(ajwerner): Track the previous stage's phase so we can know when
-		// we need to emit this op. Also, this is buggy, we need to mark the job
-		// as non-cancelable *before* we start to execute the first NonRevertible
-		// phase.
-		ops = append(ops, sb.bc.updateJobProgressOp(s.Phase > scop.PostCommitPhase))
+		ops = append(ops, bc.updateJobProgressOp(revertible))
 		return ops
 	default:
 		return nil
 	}
 }
 
-func (bc buildContext) createSchemaChangeJobOp() scop.Op {
+func (bc buildContext) createSchemaChangeJobOp(revertible bool) scop.Op {
 	return &scop.CreateSchemaChangerJob{
 		JobID:         bc.scJobIDSupplier(),
 		Statements:    bc.targetState.Statements,
 		Authorization: bc.targetState.Authorization,
 		DescriptorIDs: screl.AllTargetDescIDs(bc.targetState).Ordered(),
+		NonCancelable: !revertible,
 	}
 }
 
-func (bc buildContext) updateJobProgressOp(isNonCancellable bool) scop.Op {
+func (bc buildContext) updateJobProgressOp(revertible bool) scop.Op {
 	return &scop.UpdateSchemaChangerJob{
 		JobID:           bc.scJobIDSupplier(),
-		IsNonCancelable: isNonCancellable,
+		IsNonCancelable: !revertible,
```

Comment:
nit: would it be possible, throughout the schema changer, to only use one of these two terms, either "cancellable" or "revertible"? They mean the same thing, right?